quickwit-storage/src/bundle_storage.rs

```diff
@@ -0,0 +1,282 @@
+        self.files
+            .get(path)
+            .cloned()
+            .ok_or_else(|| io::Error::new(io::ErrorKind::NotFound, "not found in metadata").into())
+    }
+    fn exists(&self, path: &Path) -> bool {
+        self.files.contains_key(path)
+    }
+}
+
+#[async_trait]
+impl Storage for BundleStorage {
+    async fn put(&self, path: &Path, _payload: crate::PutPayload) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn copy_to_file(&self, path: &Path, _output_path: &Path) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn get_slice(&self, path: &Path, range: Range<usize>) -> crate::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        let new_range =
+            range.start + file_offsets.start as usize..range.end + file_offsets.start as usize;
+        self.storage
+            .get_slice(&self.bundle_file_name, new_range)
+            .await
+    }
+
+    async fn get_all(&self, path: &Path) -> crate::StorageResult<Bytes> {
+        let file_offsets = self.metadata.get(path)?;
+        self.storage
+            .get_slice(
+                &self.bundle_file_name,
+                file_offsets.start as usize..file_offsets.end as usize,
+            )
+            .await
+    }
+
+    async fn delete(&self, path: &Path) -> crate::StorageResult<()> {
+        Err(unsupported_operation(path))
+    }
+
+    async fn exists(&self, path: &Path) -> crate::StorageResult<bool> {
+        // also check if self.bundle_file_name exists ?
+        Ok(self.metadata.exists(path))
+    }
+
+    fn uri(&self) -> String {
+        self.storage.uri()
+    }
+}
+
+#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
+struct FileOffsets {
+    start: u64,
+    end: u64,
+}
+
+impl HasLen for BundleStorage {
+    fn len(&self) -> usize {
+        unimplemented!()
+    }
+}
+
+impl fmt::Debug for BundleStorage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "BundleStorage({:?}, files={:?})",
+            &self.bundle_file_name, self.metadata
+        )
+    }
+}
+
+fn unsupported_operation(path: &Path) -> StorageError {
+    let msg = "Unsupported operation. BundleStorage only supports async reads";
+    error!(path=?path, msg);
+    io::Error::new(io::ErrorKind::Other, format!("{}: {:?}", msg, path)).into()
+}
+
+/// BundleStorage bundles together multiple files into a single file
+/// with some metadata
+pub struct BundleStorageBuilder {
+    metadata: BundleStorageMetadata,
+    current_offset: u64,
+    /// The offset in the file where the hotcache begins. This is used to read
+    /// the hotcache and footer in a single read.
+    pub hotcache_offset: u64,
+    /// The footer offset, where the footer begins.
+    /// The footer includes the footer metadata as json encoded and the size of the footer in
+    /// bytes.
+    pub footer_offset: u64,
```

Comment:
*not a nitpick*.
I think we need to remove those `pub` and have finalize return the metadata object if you need it.

finalize could also consume self.

`fn finalize(self) -> io::Result<BundleStorageMetadata>`

Reply:
But in the creation step you don't want the BundleStorageMetadata, these offsets are only used for the single read and not used inside the bundle itself. we can return the offsets instead of pub, it's not much difference it think

Reply:
Ah yes.. We don't want the whole object.

Returning a pair of offset is not good though... Can we make it a dedicated struct with good name for the two offsets?
Shouldn't we add the file size as well?

Reply:
yes, but then we are really close to just making the two fields pub :)